architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+    data_paths: List[str]
+    data_handlers: List[DataHandlerConfig] = None
+
+@dataclass
+class DataLoaderConfig:
+    streaming: Optional[bool] = None
+
+@dataclass
+class DataConfig:
+    dataloader: Optional[DataLoaderConfig]
+    datasets: List[DatasetConfig]
+```
+
+
+The proposed design to implement support for this spec is follows,
+
+```
+class DataLoader(ABC):
+
+    tokenizer = None
+    model_name_or_path = None
+    block_size = None
+    data_config: DataConfig = None
+    data_handlers: Dict[str, Callable] = None
+
+    def __init__(self, dataconfig: DataConfig, tokenizer, model_name_or_path, block_size):
+        self.data_config = dataconfig
+        self.tokenizer = tokenizer
+        self.model_name_or_path = model_name_or_path
+        self.block_size = block_size
+        self.data_handlers = {}
+
+    def register_data_handler(self, name: str, d: Callable):
+        self.data_handlers[name] = d
+
+    @abstractmethod
+    def process_data_config(self, data_config: DataConfig):
+        pass
+```
+
+At the top level we propose to have this `class Dataloader` which is an abstract class 
+and requires functions to process the data config proposed above.
+
+We also propose a full length config verification code which preceeds the call to function 
+`Dataloader.process_data_config` as the function expects a `DataConfig` object.
+
+The data loader needs to support custom data handlers which are provided by users of the library
+or even predefined handlers which need to be registered with the data loader class using the 
+call `DataLoader.register_data_handler`.
+
+The reason to have a top level dataloader class is to ensure we have separate classes for
+hugging face and stateful data loader (fms-fsdp implementation).
+This is also` selected from the argument in the config `dataloader.type`
+
+## How are handlers provided and registered - 
+
+Data handlers are python callables which can be called on single/few samples of data and can perform
+things like applying chat template, tokenising the data, applying tools like jinja template or even
+things like encoding or decoding multi modal formats like images/audio for processing by the model.
+
+The abstract data loader class provides a way to register datahandler against a `name` which is a string.
+The data handler config `DataHandlerConfig` taken by `execute_data_handlers` represents a DAG of data handling
+routines which are to be executed on the data. 
+
+For standard HF API you can think of these as the HF Processing routines. Which could be Map/Filter/Select operations
+We implement most of the routines as map and because of this even the tokenisation of data which is done today
+in fms-hf-tuning via `tuning/utils/preprocessing_utils.py::get_preprocessed_dataset` can be retained as a data 
+handler which performs tokenization.
+
+The implementation is flexible enough for users to specify their own implementation of data handling routines
```

Comment:
users of product - i.e watsonX are not going to select what data handler to apply . Using tuning type and type of dataset , proper defaults have to be selected - watsonX is also not going to figure out , setting defaults is on us.